src/hb/shape_wasm.rs

```diff
@@ -0,0 +1,505 @@
+    _font: u32,
+    glyph: u32,
+    extents: u32,
+) -> u32 {
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+    let mut glyph_extents = hb_glyph_extents_t::default();
+    let ret = caller
+        .data()
+        .font
+        .glyph_extents(GlyphId(glyph as u16), &mut glyph_extents);
+    if ret {
+        memory.write(
+            caller.as_context_mut(),
+            extents as usize,
+            bytemuck::bytes_of(&glyph_extents),
+        );
+    }
+
+    ret as u32
+}
+
+// fn font_glyph_to_string(font: u32, glyph: u32, str: *const u8, len: u32);
+// Copies the name of the given glyph, or, if no name is available, a string of the form gXXXX into the given string.
+fn font_glyph_to_string(
+    mut caller: Caller<'_, ShapingData>,
+    _font: u32,
+    glyph: u32,
+    str: u32,
+    len: u32,
+) {
+    // len is apparently the assigned heap memory. It seems I should not allocate more than that.
+    // Should not assume I am not writing over anything.
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+    let name = caller
+        .data()
+        .font
+        .glyph_name(GlyphId(glyph as u16))
+        .map(ToOwned::to_owned)
+        .unwrap_or(format!("g{glyph:4}"));
+
+    let name = CString::new(name).unwrap();
+    let name = &name.as_bytes()[..len as usize]; // are names ever non ascii?
+
+    memory.write(caller.as_context_mut(), str as usize, name);
+}
+
+// fn font_get_glyph_h_advance(font: u32, glyph: u32) -> i32;
+// Returns the default horizontal advance for the given glyph ID the current scale and variations settings.
+fn font_get_glyph_h_advance(caller: Caller<'_, ShapingData>, _font: u32, glyph: u32) -> i32 {
+    caller.data().font.glyph_h_advance(GlyphId(glyph as u16))
+}
+
+// fn font_get_glyph_v_advance(font: u32, glyph: u32) -> i32;
+// Returns the default vertical advance for the given glyph ID the current scale and variations settings.
+fn font_get_glyph_v_advance(caller: Caller<'_, ShapingData>, _font: u32, glyph: u32) -> i32 {
+    caller.data().font.glyph_v_advance(GlyphId(glyph as u16))
+}
+
+// fn font_copy_glyph_outline(font: u32, glyph: u32, outline: *mut CGlyphOutline) -> bool;
+// Copies the outline of the given glyph ID, at current scale and variation settings, into the outline structure provided.
+fn font_copy_glyph_outline(
+    mut caller: Caller<'_, ShapingData>,
+    _font: u32,
+    glyph: u32,
+    outline: u32,
+) -> u32 {
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+    let builder = CGlyphOutline {
+        n_points: todo!(),
+        points: todo!(),
+        n_contours: todo!(),
+        contours: todo!(),
+    };
+    // also no clue what to do here
+    // let my_ol = caller.data().outline_glyph(GlyphId(glyph as u16), builder); // ??
+
+    let Ok(()) = memory.write(caller, outline as usize, todo!("builder result goes here")) else {
+        return 0;
+    };
+
```

Comment:
It's indeed interesting. I don't think the `CGlyphOutline` struct makes much sense to have in this file since the struct is meant to be operated on by the wasm module, meaning the pointers in it are wasm-sized, but this struct will have host-sized pointers. Looking at harfbuzz-wasm, the `outline` parameter represents a reference to an already existing, uninitialized `CGlyphOutline` struct in the wasm memory. But it is unclear to me how the `points` and `contours` pointers are supposed to be filled since there is no way to allocate memory in the module and we can't just randomly write somewhere.

edit: same problem as in `face_copy_table`

Reply:
The `memory.grow(1)` dance later in the file was suggested by helpful souls on the Rust Discord. I havent done it here because I was focused on getting `Calculator-Regular.ttf` working and it doesn't make use of this.

It seems the idiomatic WASM solution for this is to use `ExternRef`, but Rust compiled to `wasm` doesn't naturall support it (tho there are [crates](https://docs.rs/externref/latest/externref/)), *and* the `harfbuzz wasm examples don't make use of it. Maybe the WASM API could be changed but that's a talk for another time.

My issue for *this* function is like @RazrFalcon mentioned that `rustybuzz` doesn't keep the contours, so we might just pass the points. I'd have to test this with the Nastaliq and Ruqaa examples.

Reply:
Looking at the code in `harfbuzz-wasm`, [here](https://github.com/harfbuzz/harfbuzz/blob/e079dd20304bacd8d753eee2ceacbaa1a3978a34/src/wasm/rust/harfbuzz-wasm/src/lib.rs#L154), `n_contours` seems to be necessary, as it is used to draw the curves from which collision calculations are made.

From what I understand from the code, contours here are the indices of which points are `end_pts_of_contours`. Is there a straightforward way to get this info from `ttf-parser`?

Reply:
As said before, `ttf-parser` doesn't preserve this info. It would be really hard to do.

Also, `contours` is a `glyf` specific thing. CFF doesn't have them. No idea why WASM table relies on them to begin with.